### PR TITLE
Network progress.

### PR DESCRIPTION
**Pull request by eugeneego**

### CODE REVIEW

File: Sources/Network/Http/UrlSessionHttp.swift [hunk 1]
```diff
@@ -164,10 +167,23 @@ open class UrlSessionHttp: Http {
     // MARK: - Task
 
     private class Task: HttpTask {
+        var uploadBytes: Int64 = -1
```
Comment (bealex) [inline Sources/Network/Http/UrlSessionHttp.swift]: `-1`? Why not Optional?

File: Sources/Network/Http/UrlSessionHttp.swift [hunk 2]
```diff
@@ -164,10 +167,23 @@ open class UrlSessionHttp: Http {
...
+        var uploadTotalBytes: Int64 = -1
+        var uploadProgress: HttpProgress?
```
Comment (bealex) [inline Sources/Network/Http/UrlSessionHttp.swift]: Confusing name. I'd suggest `uploadProgressCallback` or `uploadProgressChanged`

File: Sources/Network/Http/UrlSessionHttp.swift (hunk 1 above)
Comment (eugeneego) [reply Sources/Network/Http/UrlSessionHttp.swift]: It's a default value in URLSession ¯\\_(ツ)_/¯

Comment (bealex) [reply Sources/Network/Http/UrlSessionHttp.swift]: map it to `nil`? Don't think it is a good idea to pull these things out of specific SDK.

File: Sources/Network/Http/UrlSessionHttp.swift (hunk 2 above)
Comment (eugeneego) [reply Sources/Network/Http/UrlSessionHttp.swift]: @bealex How about this
```swift
struct Progress {
    var bytes: Int64?
    var totalBytes: Int64?
    var callback: HttpProgress?
}

class Task {
    var uploadProgress: Progress
    var downloadProgress: Progress
    // ...
}
```

File: Sources/Network/Http/UrlSessionHttp.swift (hunk 1 above)
Comment (eugeneego) [reply Sources/Network/Http/UrlSessionHttp.swift]: Ok

File: Sources/Network/Http/UrlSessionHttp.swift (hunk 2 above)
Comment (bealex) [reply Sources/Network/Http/UrlSessionHttp.swift]:
```swift
struct Progress {
    var bytes: (Int64, total: Int64)?
    var callback: HttpProgress?
}
```
?

Comment (eugeneego) [reply Sources/Network/Http/UrlSessionHttp.swift]: Looks wrong. `bytes` could exist without `totalBytes`.

Comment (bealex) [reply Sources/Network/Http/UrlSessionHttp.swift]: Damn. Ok, maybe something like this then?
```swift
enum Progress {
    case unknown
    case finished(total: UInt64)
    case indeterminate(bytes: UInt64)
    case determinate(bytes: UInt64, total: UInt64)
}
```
It starts to look over engineered a little bit.

Comment (eugeneego) [reply Sources/Network/Http/UrlSessionHttp.swift]: Too complex 😉

Comment (bealex) [reply Sources/Network/Http/UrlSessionHttp.swift]: Yes, but at least noone will miss a case. :)

**Comment by bealex** [comment]
Let's see what these nils will lead to. I'm very interested to see the result.

**Review by bealex — APPROVED** [review]

**Pull request merged**